### ISSUE TITLE
Issue replicating AutoKeras StructuredDataClassifier

### ISSUE THREAD

**Issue by NullCodex** [keras-team/autokeras#1545]
I have a model that I generated using AutoKeras and I want to replicate the model so that I can construct it with keras tuner to do further hyperparameter tuning. But I am running into issues replicating the model. The model summary of the autokeras model is:
```
Model: "model"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
input_1 (InputLayer)         [(None, 11)]              0         
_________________________________________________________________
multi_category_encoding (Mul (None, 11)                0         
_________________________________________________________________
normalization (Normalization (None, 11)                23        
_________________________________________________________________
dense (Dense)                (None, 16)                192       
_________________________________________________________________
re_lu (ReLU)                 (None, 16)                0         
_________________________________________________________________
dense_1 (Dense)              (None, 32)                544       
_________________________________________________________________
re_lu_1 (ReLU)               (None, 32)                0         
_________________________________________________________________
dense_2 (Dense)              (None, 3)                 99        
_________________________________________________________________
classification_head_1 (Softm (None, 3)                 0         
=================================================================
Total params: 858
Trainable params: 835
Non-trainable params: 23
```
Layer config
```
{'batch_input_shape': (None, 11), 'dtype': 'string', 'sparse': False, 'ragged': False, 'name': 'input_1'}
{'name': 'multi_category_encoding', 'trainable': True, 'dtype': 'float32', 'encoding': ListWrapper(['int', 'int', 'int', 'int', 'int', 'int', 'int', 'int', 'int', 'int', 'int'])}
{'name': 'normalization', 'trainable': True, 'dtype': 'float32', 'axis': (-1,)}
{'name': 'dense', 'trainable': True, 'dtype': 'float32', 'units': 16, 'activation': 'linear', 'use_bias': True, 'kernel_initializer': {'class_name': 'GlorotUniform', 'config': {'seed': None}}, 'bias_initializer': {'class_name': 'Zeros', 'config': {}}, 'kernel_regularizer': None, 'bias_regularizer': None, 'activity_regularizer': None, 'kernel_constraint': None, 'bias_constraint': None}
{'name': 're_lu', 'trainable': True, 'dtype': 'float32', 'max_value': None, 'negative_slope': array(0., dtype=float32), 'threshold': array(0., dtype=float32)}
{'name': 'dense_1', 'trainable': True, 'dtype': 'float32', 'units': 32, 'activation': 'linear', 'use_bias': True, 'kernel_initializer': {'class_name': 'GlorotUniform', 'config': {'seed': None}}, 'bias_initializer': {'class_name': 'Zeros', 'config': {}}, 'kernel_regularizer': None, 'bias_regularizer': None, 'activity_regularizer': None, 'kernel_constraint': None, 'bias_constraint': None}
{'name': 're_lu_1', 'trainable': True, 'dtype': 'float32', 'max_value': None, 'negative_slope': array(0., dtype=float32), 'threshold': array(0., dtype=float32)}
{'name': 'dense_2', 'trainable': True, 'dtype': 'float32', 'units': 3, 'activation': 'linear', 'use_bias': True, 'kernel_initializer': {'class_name': 'GlorotUniform', 'config': {'seed': None}}, 'bias_initializer': {'class_name': 'Zeros', 'config': {}}, 'kernel_regularizer': None, 'bias_regularizer': None, 'activity_regularizer': None, 'kernel_constraint': None, 'bias_constraint': None}
{'name': 'classification_head_1', 'trainable': True, 'dtype': 'float32', 'axis': -1}
```
My training data is a dataframe that's converted to string type with both numerical and categorical data. Since the output is softmax i used `LabelBinarizer` to convert the target classes. 

To make sure the model was replicated properly, i used `keras.clone_model`. The optimizer and loss function follows from the original model. I was able to find the values using `model.optimizer` and `model.loss` to create a copy of the model and try training it myself. But when I tried to train it myself the accuracy does not improve despite hitting 500 epochs.

Is there something that I am missing when it comes to training the model from scratch?

**Comment by haifeng-jin** [comment]
If you just clone the model, the weights should be already trained. Did you reinitialize the weights?

**Comment by NullCodex** [comment]
I am using `keras.models.clone_model` which does not carry over the weights. I wanted to take the autokeras model as a starting point and examine the effect of adding additional layers but when i cloned the model the accuracy never improves.

**Comment by NullCodex** [comment]
To be clear, i also used model.compile with the same optimizer setting to fit the model.

**Comment by niemand-01** [comment]
Hi NullCodex,

i encountered the same problem yesterday and i found out the real problem is that we need to copy the weights in the "Multi_cetegory_encoding" layer, other weights in dense or some other layer can be learned during training, but category layer is somehow learned inside autokeras. Now i would like to know if we can train this layer or is there an equivalent layer with similiar behavior to "Multi_cetegory_encoding" layer that we can use. @haifeng-jin 

Niemand-01

**Comment by NullCodex** [comment]
@niemand-01 how are you copying over the weights? I actually ended up creating a new model by copying the layer from the base model and then appending layers. It's something like:

```
base_model = load_model()
x = base_model.layers[1](x)
```

where the index is for the multicategorical layer.

**Comment by niemand-01** [comment]
@NullCodex I suppose 
`base_model.layers[1].get_weights()` will do the trick, but i think MultiCategoryEncoder layer cant be directly integrated into a Sequential model like a normal dense layer. So i tried finding the underlying changement of MCE Layer and did a similar encoding manuelly beforehand.

**Comment by NullCodex** [comment]
actually this worked for me:
```
from tensorflow import keras
inputs = keras.layers.Input(shape=(11,), dtype='string')
x = base_model.layers[1](inputs)
x = base_model.layers[2](x)

x = keras.layers.Dense(176)(x)
x = keras.layers.ReLU()(x)
x = keras.layers.Dense(400)(x)
x = keras.layers.ReLU()(x)
x = keras.layers.Dense(464)(x)
x = keras.layers.ReLU()(x)
x = keras.layers.Dense(3)(x)
x = keras.layers.Softmax()(x)

layer_3 = keras.Model(inputs, x)
```

i was able to get a 90+ accuracy on my task with it

**Comment by hrushiwhizhack** [comment]
> actually this worked for me:
> 
> ```
> from tensorflow import keras
> inputs = keras.layers.Input(shape=(11,), dtype='string')
> x = base_model.layers[1](inputs)
> x = base_model.layers[2](x)
> 
> x = keras.layers.Dense(176)(x)
> x = keras.layers.ReLU()(x)
> x = keras.layers.Dense(400)(x)
> x = keras.layers.ReLU()(x)
> x = keras.layers.Dense(464)(x)
> x = keras.layers.ReLU()(x)
> x = keras.layers.Dense(3)(x)
> x = keras.layers.Softmax()(x)
> 
> layer_3 = keras.Model(inputs, x)
> ```
> 
> i was able to get a 90+ accuracy on my task with it

Hi 
Am facing the same issue ... & am not able to solve this thing can you please help me.
